### PR TITLE
Remove in-place allocation and revert to separate methods for zeroed allocations

### LINKED ISSUE #58: Separate methods v.s. "mode" parameters
https://github.com/rust-lang/wg-allocators/issues/41 has split `realloc` into separate `grow` and `shrink` methods. However as of this writing some methods take `AllocInit` or `ReallocPlacement` parameters that are effectively boolean mode switches.

Part of the reasoning for separate methods is that most callers are always in one of the two cases, so separating saves a branch in the allocator implementation. The same seems to apply to `AllocInit` and `ReallocPlacement`.

The current design with 5 methods seems inconsistent to me. I think the trait should pick one principle among easier/nicer-looking API or single-purpose methods. This would give either 4 methods:

* `alloc`
* `dealloc`
* `realloc`
* `by_ref`

Or 10 methods:

* `alloc`
* `alloc_zeroed`
* `dealloc`
* `grow`
* `grow_zeroed`
* `grow_in_place`
* `grow_in_place_zeroed`
* `shrink`
* `shrink_in_place`
* `by_ref`

I tend to favor the latter. The reason against it given in https://github.com/rust-lang/wg-allocators/issues/44 is that this "blows up the trait". But I feel 10 methods is not that bad, especially given that they are not 10 completely different behaviors that need to be learned independently, but follow a pattern that’s reflected in their names.

### PR DESCRIPTION
**Pull request by TimDiekmann**
closes rust-lang/wg-allocators#58

### CODE REVIEW

**Review by Amanieu — COMMENTED** [review]
Are you also going to remove `MemoryBlock` in this PR or will you be doing it later in a separate PR?

**Comment by TimDiekmann** [comment]
> Are you also going to remove `MemoryBlock` in this PR or will you be doing it later in a separate PR?

I'll remove `MemoryBlock` in another PR to keep things clean. I have the changes already made locally and it's going to be a rather small PR.

File: library/core/src/alloc/mod.rs [hunk 1]
```diff
@@ -213,31 +155,87 @@ pub unsafe trait AllocRef {
...
-    /// allocation referenced by `ptr` to fit the new layout. If the [`placement`] is
-    /// [`InPlace`], the returned pointer is guaranteed to be the same as the passed `ptr`.
-    ///
-    /// If [`MayMove`] is used then ownership of the memory block referenced by `ptr`
-    /// is transferred to this allocator. The memory may or may not be freed, and should be
-    /// considered unusable (unless of course it is transferred back to the caller again via the
-    /// return value of this method).
-    ///
+    /// allocation referenced by `ptr` to fit the new layout.
+    ///~
```
Comment (Amanieu) [inline library/core/src/alloc/mod.rs:159]: Stray `~`.

File: library/core/src/alloc/mod.rs [hunk 2]
```diff
@@ -177,10 +99,25 @@ pub unsafe trait AllocRef {
...
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    fn alloc(&mut self, layout: Layout) -> Result<MemoryBlock, AllocErr>;
+
+    /// Behaves like `alloc`, but also ensures that the contents are set to zero before being returned.
```
Comment (Amanieu) [inline library/core/src/alloc/mod.rs:120]:
```suggestion
    /// Behaves like `alloc`, but also ensures that the returned memory is zero-initialized.
```

File: library/core/src/alloc/mod.rs [hunk 3]
```diff
@@ -213,31 +155,87 @@ pub unsafe trait AllocRef {
...
+    /// # Safety
+    ///
+    /// * `ptr` must denote a block of memory [*currently allocated*] via this allocator,
+    /// * `layout` must [*fit*] that block of memory (The `new_size` argument need not fit it.),
+    // We can't require that `new_size` is strictly greater than `memory.size` because of ZSTs.
+    // An alternative would be
+    // * `new_size must be strictly greater than `memory.size` or both are zero
+    /// * `new_size` must be greater than or equal to `layout.size()`, and
+    /// * `new_size`, when rounded up to the nearest multiple of `layout.align()`, must not overflow
+    ///   (i.e., the rounded value must be less than or equal to `usize::MAX`).
```
Comment (Amanieu) [inline library/core/src/alloc/mod.rs:172]: There is a mix of `//` and `///` comments here. I think the second part is not meant to be part of the docs.

File: library/core/src/alloc/mod.rs [hunk 4]
```diff
@@ -265,45 +263,39 @@ pub unsafe trait AllocRef {
...
-                    // SAFETY: the caller must ensure that the `new_size` does not overflow.
-                    // `layout.align()` comes from a `Layout` and is thus guaranteed to be valid for a Layout.
-                    // The caller must ensure that `new_size` is greater than zero.
-                    unsafe { Layout::from_size_align_unchecked(new_size, layout.align()) };
-                let new_memory = self.alloc(new_layout, init)?;
+        let new_layout =
+            // SAFETY: the caller must ensure that the `new_size` does not overflow.
+            // `layout.align()` comes from a `Layout` and is thus guaranteed to be valid for a Layout.
+            // The caller must ensure that `new_size` is greater than or equal to zero. If it's equal
+            // to zero, it's catched beforehand.
```
Comment (Amanieu) [inline library/core/src/alloc/mod.rs:286]:
```suggestion
            // to zero, it's caught beforehand.
```

File: library/core/src/alloc/mod.rs (hunk 3 above)
Comment (TimDiekmann) [reply library/core/src/alloc/mod.rs:172]: The second parts belongs to the safety section whereas the first part only describes, why we have to require "greater than **or equal**". I can rearrange the non-doc comments a bit to clarify, that this is intended.

File: library/core/src/alloc/mod.rs (hunk 1 above)
Comment (TimDiekmann) [reply library/core/src/alloc/mod.rs:159]:
```suggestion
    ///
```

**Pull request merged**